store/store.go

```diff
@@ -233,6 +240,125 @@ func (s *SnapUbuntuStoreRepository) checkStoreResponse(resp *http.Response) {
+    "purchase_id": "1",
+    "refundable_until": null,
+    "state": "Complete"
+  },
+  {
+    "open_id": "https://login.staging.ubuntu.com/+id/open_id",
+    "package_name": "com.ubuntu.developer.dev.otherapp",
+    "refundable_until": "2015-07-17 11:33:29",
+    "state": "Complete"
+  }
+]
+*/
+type purchase struct {
+	OpenID          string `json:"open_id"`
+	SnapID          string `json:"snap_id"`
+	RefundableUntil string `json:"refundable_until"`
+	State           string `json:"state"`
+	ItemSKU         string `json:"item_sku,omitempty"`
+	PurchaseID      string `json:"purchase_id,omitempty"`
+	RedirectTo      string `json:"redirect_to,omitempty"`
+}
+
+type purchaseList []*purchase
+
+func (s *SnapUbuntuStoreRepository) getPurchasesFromURL(url *url.URL, channel string, auther Authenticator) (purchaseList, error) {
+	if auther == nil {
+		return nil, fmt.Errorf("cannot obtain known purchases from store: no authentication credentials provided")
+	}
+
+	req, err := http.NewRequest("GET", url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	s.applyUbuntuStoreHeaders(req, "", auther)
+	req.Header.Set("X-Ubuntu-Device-Channel", channel)
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	var purchases purchaseList
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		dec := json.NewDecoder(resp.Body)
+		if err := dec.Decode(&purchases); err != nil {
+			return nil, fmt.Errorf("cannot decode known purchases from store: %v", err)
+		}
+	case http.StatusNotFound:
+		// the store returns 404 for snaps that have no purchases, so just return an empty list
+		break
+	case http.StatusUnauthorized:
+		// TODO handle token expiry and refresh
+		return nil, ErrInvalidCredentials
+	default:
+		return nil, fmt.Errorf("cannot obtain known purchases from store: server returned %v code", resp.StatusCode)
+	}
+
+	return purchases, nil
+}
+
+// getPurchases retreives the user's purchases for a specific package, including in-app purchases, as a list
+func (s *SnapUbuntuStoreRepository) getPurchases(snapID, channel string, auther Authenticator) (purchaseList, error) {
+	purchasesURL, err := s.purchasesURI.Parse(snapID + "/")
+	if err != nil {
+		return nil, err
+	}
+
+	q := purchasesURL.Query()
+	q.Set("include_item_purchases", "true")
+	purchasesURL.RawQuery = q.Encode()
+
+	return s.getPurchasesFromURL(purchasesURL, channel, auther)
+}
+
+// getAllPurchases retreives all the user's purchases, including in-app purchases, as a map indexed by Snap ID
+func (s *SnapUbuntuStoreRepository) getAllPurchases(channel string, auther Authenticator) (map[string]purchaseList, error) {
+	purchases, err := s.getPurchasesFromURL(s.purchasesURI, channel, auther)
+	if err != nil {
+		return nil, err
+	}
+
+	// Index it all in a multimap
+	purchasesByName := make(map[string]purchaseList)
+	for _, purchase := range purchases {
+		purchasesByName[purchase.SnapID] = append(purchasesByName[purchase.SnapID], purchase)
```

Comment:
later on, I want to associate the purchases against a bunch of the snaps, and it makes the code easier to read if I can look up the purchases for a snap by doing purchases["snap_id"], instead of searching the slice each time